### ISSUE TITLE
Multiple Firefox instances failing with NS_ERROR_SOCKET_ADDRESS_IN_USE

### ISSUE THREAD

**Issue by MoCoTester** [mozilla/geckodriver#99]
While attempting to run a Selenium test using multiple Firefox browser instances in parallel, only one browser instance is being used as expected while the others are failing.

For the browser instances that are failing, they will open but nothing happens on them.

The errors I am getting are shown below in the Console Output.

I have verified that this same test works fine using ChromeDriver (v2.9) to run multiple Chrome (v51.0.2704.84 64-bit) browser instances.

To run this project on your local machine, please update the path references in the DriverFactory.java file for your Chrome and Firefox drivers.

To run the test in Chrome, uncomment the chrome line and comment the firefox line in the Parameterized.Parameters section of the TestBase.java file.

Virtual Box Version: 5.0.16 r 105871
Operating System: Ubuntu 16.04 64 bit desktop
Java Version: openjdk 1.8.0_91
IDE: Eclipse Mars.2
Selenium Version: 2.53.0
Geckodriver Version: v0.8.0
Maven Version: 3.3.9
Language: Java
Firefox Version: 47.0
JUnit Version: 4.12

Here are the files that I am using in my project.

**DriverFactory.java**
```
package parallel;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.firefox.MarionetteDriver;

class DriverFactory {
    
    static WebDriver createInstance(String browserName) {
        
        WebDriver driver = null;
        
        if (browserName.toLowerCase().contains("firefox")) {
                System.setProperty("webdriver.gecko.driver", "/home/<user>/Downloads/wires");
                driver = new MarionetteDriver();
                return driver;
        }
        if (browserName.toLowerCase().contains("chrome")) {
                System.setProperty("webdriver.chrome.driver", "/home/<user>/Downloads/chromedriver");
                driver = new ChromeDriver();
                return driver;
        }
        return driver;
    }
    
}
```

**DriverManager.java**
```
package parallel;

import org.openqa.selenium.WebDriver;

public class DriverManager {
    
    private static ThreadLocal<WebDriver> webDriver = new ThreadLocal<WebDriver>();
    
    public static WebDriver getDriver() {
        return webDriver.get();
    }
    
    static void setWebDriver(WebDriver driver) {
        webDriver.set(driver);
    }
    
    public static void removeDriver() {
        if (webDriver.get()!=null) {
            webDriver.get().quit();
            webDriver.remove();
        }
    }

}
```

**Parallelized.java**
```
package parallel;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import org.junit.runners.Parameterized;
import org.junit.runners.model.RunnerScheduler;

public class Parallelized extends Parameterized
{
    
    private static class ThreadPoolScheduler implements RunnerScheduler
    {
        private ExecutorService executor;
        
        public ThreadPoolScheduler()
        {
            String threads = System.getProperty("junit.parallel.threads", "16");
            int numThreads = Integer.parseInt(threads);
            executor = Executors.newFixedThreadPool(numThreads);
        }
        
        @Override
        public void finished()
        {
            executor.shutdown();
            try
            {
                executor.awaitTermination(10, TimeUnit.MINUTES);
            }
            catch (InterruptedException exc)
            {
                throw new RuntimeException(exc);
            }
        }

        @Override
        public void schedule(Runnable childStatement)
        {
            executor.submit(childStatement);
        }
    }

    public Parallelized(Class klass) throws Throwable
    {
        super(klass);
        setScheduler(new ThreadPoolScheduler());
    }
}
```

**TestBase.java**
```
package parallel;

import java.util.LinkedList;

import org.junit.After;
import org.junit.Before;
import org.junit.runners.Parameterized;
import org.openqa.selenium.WebDriver;

public class TestBase {

    @Before
    public void driverInit() {
        WebDriver driver = DriverFactory.createInstance(currentDriverType);
        DriverManager.setWebDriver(driver);
    }

    @Parameterized.Parameter
    public String currentDriverType;
    
    @Parameterized.Parameters(name= "{0}")
      public static LinkedList<String[]> getEnvironments() throws Exception {
        LinkedList<String[]> env = new LinkedList<String[]>();
        //env.add(new String[]{"chrome"});
        env.add(new String[]{"firefox"});
        return env;
      }
      
      @After
      public void driverTearDown() {
          DriverManager.removeDriver();
      }
    
}
```

**ParallelSeleniumTest.java**
```
package parallel;

import org.junit.Test;
import org.junit.runner.RunWith;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.CoreMatchers.containsString;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

@RunWith(Parallelized.class)
public class ParallelSeleniumTest extends TestBase {

    @Test
    public void searchForGoogle() throws InterruptedException {
        invokeBrowser("http://www.google.com");
        WebElement searchBox = DriverManager.getDriver().findElement(By.name("q"));
        searchBox.sendKeys("dogs");
        searchBox.submit();
        Thread.sleep(3000);
        assertThat(DriverManager.getDriver().getTitle(), containsString("dogs"));
    }

    @Test
    public void searchForGitHub() throws InterruptedException {
        invokeBrowser("http://www.google.com");
        WebElement searchBox = DriverManager.getDriver().findElement(By.name("q"));
        searchBox.sendKeys("cats");
        searchBox.submit();
        Thread.sleep(3000);
        assertThat(DriverManager.getDriver().getTitle(), containsString("cats"));
    }

    @Test
    public void searchForStackoverflow() throws InterruptedException {
        invokeBrowser("http://www.google.com");
        WebElement searchBox = DriverManager.getDriver().findElement(By.name("q"));
        searchBox.sendKeys("birds");
        searchBox.submit();
        Thread.sleep(3000);
        assertThat(DriverManager.getDriver().getTitle(), containsString("birds"));
    }

    private void invokeBrowser(String url) {
        System.out.println(currentDriverType + " Thread ID = " + Thread.currentThread().getId());
        System.out.println("Hashcode of " + currentDriverType + " webDriver instance = " + DriverManager.getDriver().hashCode());
        DriverManager.getDriver().get(url);
    }

}
```

**pom.xml**
```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.<company>.selenium</groupId>
  <artifactId>parallel</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  
    <dependencies>
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.12</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.seleniumhq.selenium</groupId>
      <artifactId>selenium-java</artifactId>
      <version>2.53.0</version>
      <scope>test</scope>
    </dependency>
  </dependencies>
  
  <build>
  	<plugins>
		<plugin>
  			<groupId>org.apache.maven.plugins</groupId>
  			<artifactId>maven-surefire-plugin</artifactId>
  			<version>2.19.1</version>
  			<configuration>
  				<parallel>all</parallel>
  				<threadCount>3</threadCount>
  			</configuration>
  		</plugin>
  	</plugins>
  </build>
</project>
```

**Console Output**
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building parallel 0.0.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ parallel ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] Copying 0 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ parallel ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ parallel ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] Copying 0 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ parallel ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-surefire-plugin:2.19.1:test (default-test) @ parallel ---

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running [firefox]
1466172953730	Marionette	DEBUG	Marionette enabled via build flag and pref
1466172954364	Marionette	ERROR	Error on starting server: [Exception... "Component returned failure code: 0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE) [nsIServerSocket.initSpecialConnection]"  nsresult: "0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE)"  location: "JS frame :: chrome://marionette/content/server.js :: MarionetteServer.prototype.start :: line 105"  data: no][Exception... "Component returned failure code: 0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE) [nsIServerSocket.initSpecialConnection]"  nsresult: "0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE)"  location: "JS frame :: chrome://marionette/content/server.js :: MarionetteServer.prototype.start :: line 105"  data: no]MarionetteServer.prototype.start@chrome://marionette/content/server.js:105:19MarionetteComponent.prototype.init@resource://gre/components/marionettecomponent.js:184:5MarionetteComponent.prototype.observe@resource://gre/components/marionettecomponent.js:127:7
1466172955530	Marionette	DEBUG	Marionette enabled via command-line flagfirefox Thread ID = 12
Hashcode of firefox webDriver instance = 1437605705
JavaScript warning: https://normandy.cdn.mozilla.net/static/js/bundles/selfrepair-7575f6d27445b45bcffb.js, line 9825: mutating the [[Prototype]] of an object will cause your code to run very slowly; instead create the object with the correct initial [[Prototype]] value using Object.create
1466172953619	Marionette	DEBUG	Marionette enabled via build flag and pref
1466172954387	Marionette	ERROR	Error on starting server: [Exception... "Component returned failure code: 0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE) [nsIServerSocket.initSpecialConnection]"  nsresult: "0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE)"  location: "JS frame :: chrome://marionette/content/server.js :: MarionetteServer.prototype.start :: line 105"  data: no][Exception... "Component returned failure code: 0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE) [nsIServerSocket.initSpecialConnection]"  nsresult: "0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE)"  location: "JS frame :: chrome://marionette/content/server.js :: MarionetteServer.prototype.start :: line 105"  data: no]MarionetteServer.prototype.start@chrome://marionette/content/server.js:105:19MarionetteComponent.prototype.init@resource://gre/components/marionettecomponent.js:184:5MarionetteComponent.prototype.observe@resource://gre/components/marionettecomponent.js:127:7
1466172955685	Marionette	DEBUG	Marionette enabled via command-line flagJavaScript warning: https://normandy.cdn.mozilla.net/static/js/bundles/selfrepair-7575f6d27445b45bcffb.js, line 9825: mutating the [[Prototype]] of an object will cause your code to run very slowly; instead create the object with the correct initial [[Prototype]] value using Object.create
1466172953695	Marionette	DEBUG	Marionette enabled via build flag and pref
1466172954276	Marionette	INFO	Listening on port 2828
1466172955455	Marionette	DEBUG	Marionette enabled via command-line flag
1466172955655	Marionette	DEBUG	Accepted connection conn0 from 127.0.0.1:56206
1466172955664	Marionette	DEBUG	Accepted connection conn1 from 127.0.0.1:56208
1466172955667	Marionette	DEBUG	Accepted connection conn2 from 127.0.0.1:56210
1466172955822	Marionette	TRACE	conn0 -> [0,1,"newSession",{"capabilities":{"desiredCapabilities":{"browserName":"firefox","platform":"ANY","version":""},"requiredCapabilities":{}},"sessionId":null}]
1466172955827	Marionette	CONFIG	Changing capabilities: {"browserName":"Firefox","browserVersion":"47.0","platformName":"Linux","platformVersion":"4.4.0-24-generic","specificationLevel":0,"raisesAccessibilityExceptions":false,"rotatable":false,"acceptSslCerts":false,"takesElementScreenshot":true,"takesScreenshot":true,"proxy":{},"platform":"LINUX","XULappId":"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}","appBuildId":"20160606113944","device":"desktop","version":""}
1466172955829	Marionette	TRACE	conn1 -> [0,1,"newSession",{"capabilities":{"desiredCapabilities":{"browserName":"firefox","platform":"ANY","version":""},"requiredCapabilities":{}},"sessionId":null}]
1466172955830	Marionette	CONFIG	Changing capabilities: {"browserName":"Firefox","browserVersion":"47.0","platformName":"Linux","platformVersion":"4.4.0-24-generic","specificationLevel":0,"raisesAccessibilityExceptions":false,"rotatable":false,"acceptSslCerts":false,"takesElementScreenshot":true,"takesScreenshot":true,"proxy":{},"platform":"LINUX","XULappId":"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}","appBuildId":"20160606113944","device":"desktop","version":""}
1466172955844	Marionette	TRACE	conn2 -> [0,1,"newSession",{"capabilities":{"desiredCapabilities":{"browserName":"firefox","platform":"ANY","version":""},"requiredCapabilities":{}},"sessionId":null}]
1466172955845	Marionette	CONFIG	Changing capabilities: {"browserName":"Firefox","browserVersion":"47.0","platformName":"Linux","platformVersion":"4.4.0-24-generic","specificationLevel":0,"raisesAccessibilityExceptions":false,"rotatable":false,"acceptSslCerts":false,"takesElementScreenshot":true,"takesScreenshot":true,"proxy":{},"platform":"LINUX","XULappId":"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}","appBuildId":"20160606113944","device":"desktop","version":""}
1466172958373	Marionette	DEBUG	loaded listener.jsJavaScript error: chrome://marionette/content/driver.js, line 387: TypeError: this.curBrowser is nullJavaScript error: chrome://marionette/content/driver.js, line 387: TypeError: this.curBrowser is null
1466172958394	Marionette	DEBUG	loaded listener.jsJavaScript error: chrome://marionette/content/driver.js, line 387: TypeError: this.curBrowser is nullJavaScript error: chrome://marionette/content/driver.js, line 387: TypeError: this.curBrowser is nullJavaScript error: , line 0: WebDriverError: Session already runningJavaScript error: , line 0: WebDriverError: Session already running
1466172959004	Marionette	TRACE	conn2 <- [1,1,null,{"sessionId":"5c327c60-bef3-4b9c-859d-97ec4e621be3","capabilities":{"browserName":"Firefox","browserVersion":"47.0","platformName":"Linux","platformVersion":"4.4.0-24-generic","specificationLevel":0,"raisesAccessibilityExceptions":false,"rotatable":false,"acceptSslCerts":false,"takesElementScreenshot":true,"takesScreenshot":true,"proxy":{},"platform":"LINUX","XULappId":"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}","appBuildId":"20160606113944","device":"desktop","version":"","command_id":1}}]
1466172959181	Marionette	TRACE	conn2 -> [0,2,"get",{"url":"http://www.google.com"}]
1466172962346	Marionette	TRACE	conn2 <- [1,2,null,{}]
1466172962455	Marionette	TRACE	conn2 -> [0,3,"findElement",{"using":"css selector","value":"*[name='q']"}]
1466172962514	Marionette	TRACE	conn2 <- [1,3,null,{"value":{"element-6066-11e4-a52e-4f735466cecf":"7c58d627-6a59-46be-8285-c4405b9261ec","ELEMENT":"7c58d627-6a59-46be-8285-c4405b9261ec"}}]
1466172962622	Marionette	TRACE	conn2 -> [0,4,"sendKeysToElement",{"id":"7c58d627-6a59-46be-8285-c4405b9261ec","value":["b","i","r","d","s"]}]
1466172962651	Marionette	DEBUG	Element does not have an accessible object: id: lst-ib, tagName: INPUT, className: gsfi lst-d-f
1466172962860	Marionette	TRACE	conn2 <- [1,4,null,{}](firefox:7902): GLib-GObject-CRITICAL **: g_type_add_interface_static: assertion 'g_type_parent (interface_type) == G_TYPE_INTERFACE' failed
1466172963048	Marionette	TRACE	conn2 -> [0,5,"findElement",{"element":"7c58d627-6a59-46be-8285-c4405b9261ec","using":"xpath","value":"./ancestor-or-self::form"}](firefox:7902): GLib-GObject-CRITICAL **: g_type_add_interface_static: assertion 'g_type_parent (interface_type) == G_TYPE_INTERFACE' failed(firefox:7902): GLib-GObject-CRITICAL **: g_type_add_interface_static: assertion 'g_type_parent (interface_type) == G_TYPE_INTERFACE' failed(firefox:7902): GLib-GObject-CRITICAL **: g_type_add_interface_static: assertion 'g_type_parent (interface_type) == G_TYPE_INTERFACE' failed
1466172963104	Marionette	TRACE	conn2 <- [1,5,null,{"value":{"element-6066-11e4-a52e-4f735466cecf":"07786632-2dec-4dee-b4d3-2366a9dc5d5e","ELEMENT":"07786632-2dec-4dee-b4d3-2366a9dc5d5e"}}]
1466172963129	Marionette	TRACE	conn2 -> [0,6,"executeScript",{"args":[{"ELEMENT":"07786632-2dec-4dee-b4d3-2366a9dc5d5e","element-6066-11e4-a52e-4f735466cecf":"07786632-2dec-4dee-b4d3-2366a9dc5d5e"}],"newSandbox":false,"script":"var e = arguments[0].ownerDocument.createEvent('Event');e.initEvent('submit', true, true);if (arguments[0].dispatchEvent(e)) { arguments[0].submit() }","scriptTimeout":null,"specialPowers":false}]
1466172963201	Marionette	TRACE	conn2 <- [1,6,null,{"value":null}](firefox:7902): GLib-GObject-CRITICAL **: g_type_add_interface_static: assertion 'g_type_parent (interface_type) == G_TYPE_INTERFACE' failedJavaScript warning: https://normandy.cdn.mozilla.net/static/js/bundles/selfrepair-7575f6d27445b45bcffb.js, line 9825: mutating the [[Prototype]] of an object will cause your code to run very slowly; instead create the object with the correct initial [[Prototype]] value using Object.create
1466172966246	Marionette	TRACE	conn2 -> [0,7,"getTitle",{}]
1466172966256	Marionette	TRACE	conn2 <- [1,7,null,{"value":"birds - Google Search"}]
1466172966279	Marionette	TRACE	conn2 -> [0,8,"quitApplication",{"flags":["eForceQuit"]}]
1466172966280	Marionette	TRACE	conn2 <- [1,8,null,{}]
1466172966617	Marionette	DEBUG	Closed connection conn2Tests run: 3, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 17.49 sec <<< FAILURE! - in [firefox]
searchForGoogle[firefox](parallel.ParallelSeleniumTest)  Time elapsed: 17.488 sec  <<< ERROR!
org.openqa.selenium.WebDriverException: 
Failed to decode response from marionette (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 16.97 seconds
Build info: version: '2.53.0', revision: '35ae25b1534ae328c771e0856c93e187490ca824', time: '2016-03-15 10:43:46'
System info: host: '<user>-VirtualBox', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-24-generic', java.version: '1.8.0_91'
Driver info: org.openqa.selenium.firefox.MarionetteDriver

searchForGitHub[firefox](parallel.ParallelSeleniumTest)  Time elapsed: 17.484 sec  <<< ERROR!
org.openqa.selenium.WebDriverException: 
Failed to decode response from marionette (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 16.94 seconds
Build info: version: '2.53.0', revision: '35ae25b1534ae328c771e0856c93e187490ca824', time: '2016-03-15 10:43:46'
System info: host: '<user>-VirtualBox', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-24-generic', java.version: '1.8.0_91'
Driver info: org.openqa.selenium.firefox.MarionetteDriver

Results :

Tests in error: 
  ParallelSeleniumTest>TestBase.driverInit:14 » WebDriver Failed to decode respo...
  ParallelSeleniumTest>TestBase.driverInit:14 » WebDriver Failed to decode respo...

Tests run: 3, Failures: 0, Errors: 2, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 20.826 s
[INFO] Finished at: 2016-06-17T10:16:07-04:00
[INFO] Final Memory: 15M/170M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project parallel: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/<user>/workspace/parallel/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

**Comment by twalpole** [comment]
This is because no --marionette-port option is passed to geckodriver - which ends all instances of geckodriver launch firefox passing the same desired default port (2828).  The first firefox instance binds to that port, future instances can't and all the geckodriver instances end up connecting to the first firefox instance - which produces all sorts of unpredictable behavior.

**Comment by andreastt** [comment]
I think a reasonable short-term solution is to do what the other drivers are doing and ask Marionette to bind to a randomised, free port generated by geckodriver. Currently it uses 2828 as the default for all instances it spawns of Firefox.

Since Marionette unfortunately does not yet have an out-of-band way of communicating the port back to the client (geckodriver), this is inherently racy but we can improve the situation in the future with one of the proposals from [bug 1240830](https://bugzilla.mozilla.org/show_bug.cgi?id=1240830#c3).

**Comment by jgraham** [comment]
I believe this will be fixed by #112 Please reopen if the firthcoming 0.9.0 release (or latest master) doesn't fix your issue.
